src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -0,0 +1,805 @@
+  Ref<InternalNode> this_ref = this;
+  return (found == tracked_child_nodes.end()
+    ? (logger().trace("OTree::Internal: load child untracked at {:#x}, pos({}), level={}",
+                      child_addr, position, level() - 1),
+       Node::load(c, child_addr, level_tail
+       ).safe_then([this, position] (auto child) {
+         child->as_child(position, this);
+         return child;
+       }))
+    : (logger().trace("OTree::Internal: load child tracked at {:#x}, pos({}), level={}",
+                      child_addr, position, level() - 1),
+       node_ertr::make_ready_future<Ref<Node>>(found->second))
+  ).safe_then([this_ref, this, position, child_addr] (auto child) {
+    assert(child_addr == child->impl->laddr());
+    assert(position == child->parent_info().position);
+    validate_child(*child);
+    return child;
+  });
+}
+
+void InternalNode::track_insert(
+      const search_position_t& insert_pos, match_stage_t insert_stage,
+      Ref<Node> insert_child, Ref<Node> nxt_child) {
+  // update tracks
+  auto pos_upper_bound = insert_pos;
+  pos_upper_bound.index_by_stage(insert_stage) = INDEX_UPPER_BOUND;
+  auto first = tracked_child_nodes.lower_bound(insert_pos);
+  auto last = tracked_child_nodes.lower_bound(pos_upper_bound);
+  std::vector<Node*> nodes;
+  std::for_each(first, last, [&nodes](auto& kv) {
+    nodes.push_back(kv.second);
+  });
+  tracked_child_nodes.erase(first, last);
+  for (auto& node : nodes) {
+    auto _pos = node->parent_info().position;
+    assert(!_pos.is_end());
+    ++_pos.index_by_stage(insert_stage);
+    node->as_child(_pos, this);
+  }
+  // track insert
+  insert_child->as_child(insert_pos, this);
+
+#ifndef NDEBUG
+  // validate left_child is before right_child
+  if (nxt_child) {
+    auto iter = tracked_child_nodes.find(insert_pos);
+    ++iter;
+    assert(iter->second == nxt_child);
+  }
+#endif
+}
+
+void InternalNode::replace_track(
+    const search_position_t& position, Ref<Node> new_child, Ref<Node> old_child) {
+  assert(tracked_child_nodes[position] == old_child);
+  tracked_child_nodes.erase(position);
+  new_child->as_child(position, this);
+  assert(tracked_child_nodes[position] == new_child);
+}
+
+void InternalNode::track_split(
+    const search_position_t& split_pos, Ref<InternalNode> right_node) {
+  auto first = tracked_child_nodes.lower_bound(split_pos);
+  auto iter = first;
+  while (iter != tracked_child_nodes.end()) {
+    search_position_t new_pos = iter->first;
+    new_pos -= split_pos;
+    iter->second->as_child<false>(new_pos, right_node);
+    ++iter;
+  }
+  tracked_child_nodes.erase(first, tracked_child_nodes.end());
+}
+
+void InternalNode::validate_child(const Node& child) const {
+#ifndef NDEBUG
+  assert(impl->level() - 1 == child.impl->level());
+  assert(this == child.parent_info().ptr);
+  auto& child_pos = child.parent_info().position;
+  assert(impl->get_p_value(child_pos)->value == child.impl->laddr());
+  if (child_pos.is_end()) {
```

Comment:
I'd sort of like any usage of 'end' to denote one-past-the-end with 'last' denoting last valid.